src/crud_ops.rs

```diff
@@ -143,6 +225,51 @@ impl Client {
         counter!(metrics::total_delete_ops).increment(1);
         with_retries!(self, self.delete_impl(path).await)
     }
+
+    async fn bulk_delete_impl(&self, paths: &Vec<Path>) -> crate::Result<Vec<(Path, crate::Error)>> {
+        let stream = stream::iter(paths.iter().map(|path| Ok(path.clone()))).boxed();
+        let indexed_results = self.store.delete_stream(stream)
+            .enumerate()
+            .filter_map(|(index, result)| async move {
+                Some((index, result))
+            }).collect::<Vec<(usize, Result<Path, object_store::Error>)>>().await;
+        // We need to count the number of callbacks called to determine if we need to raise an error
+        // if some paths were not processed at all.
+        let callbacks_called = indexed_results.len();
+        let bulk_failed_entries = indexed_results.into_iter().filter_map(|(index, result)| {
+            match result {
+                Ok(_) => {
+                    None
+                },
+                Err(e) => match e {
+                    // We treat not found as success because AWS S3 does not return an error
+                    // if the object does not exist
+                    object_store::Error::NotFound { .. } => {
+                        None
+                    },
+                    _ => {
+                        Some((paths[index].clone(), e.into()))
+                    }
+                },
+            }
+        }).collect::<Vec<(Path, crate::Error)>>();
+
+        // Rail guard to catch generic errors
+        if callbacks_called < paths.len() {
```

Comment:
I think `callbacks_called` now deserves another name, maybe just using `results.len()` is good enough?

Reply:
ended up using `num_results`